src/main/scala/com/linkedin/feathr/offline/job/FeatureTransformation.scala

```diff
@@ -860,8 +882,18 @@ private[offline] object FeatureTransformation {
     val (directTransformAnchorGroup, incrementalTransformAnchorGroup) =
       groupAggregationFeatures(source, AnchorFeatureGroups(anchorsWithSameSource, allRequestedFeatures), incrementalAggContext)
 
+    // Obtain the preprocessed DataFrame by Pyspark for this anchor to replace the origin source DataFrame.
+    // At this point, we have make the anchors that have preprocessing UDF unique so it will be only one anchor if
+    // the anchor has preprocessing UDF.
+    // Then we use the feature names separated by comma as the key to get the preprocessed DataFrame from Pyspark.
+    // If there are multiple features in an anchor, there will be multiple anchors in ths anchorsWithSameSource
+    // so we need to merge the duplicates via set, for example [f1, f1, f2, f2] => [f1, f2]
+    val allAnchorsFeatures = anchorsWithSameSource.flatMap(_.featureAnchor.features).toSet.toList.sorted
```

Comment:
Call a shared function instead?

Reply:
it's not too complex so it might be a bit overkill to call another function?